src/StackExchange.Redis/Configuration/DefaultOptionsProvider.cs

```diff
@@ -0,0 +1,220 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Reflection;
+using System.Threading.Tasks;
+
+namespace StackExchange.Redis.Configuration
+{
+    /// <summary>
+    /// A defaults providers for <see cref="ConfigurationOptions"/>.
+    /// This providers defaults not explicitly specified and is present to be inherited by environments that want to provide
+    /// better defaults for their use case, e.g. in a single wrapper library used many places.
+    /// </summary>
+    /// <remarks>
+    /// Why not just have a default <see cref="ConfigurationOptions"/> instance? Good question!
+    /// Since we null coalesce down to the defaults, there's an inherent pit-of-failure with that approach of <see cref="StackOverflowException"/>.
+    /// If you forget anything or if someone creates a provider nulling these out...kaboom.
+    /// </remarks>
+    public class DefaultOptionsProvider
+    {
+        private static readonly List<DefaultOptionsProvider> BuiltInProviders = new()
+        {
+            new AzureOptionsProvider()
+        };
+        private static LinkedList<DefaultOptionsProvider> KnownProviders { get; } = new (BuiltInProviders);
+
+        /// <summary>
+        /// Adds a provider to match endpoints against. The last provider added has the highest priority.
+        /// If you want your provider to match everything, implement <see cref="IsMatch(EndPoint)"/> as <c>return true;</c>.
+        /// </summary>
+        /// <param name="provider">The provider to add.</param>
+        public static void AddProvider(DefaultOptionsProvider provider) => KnownProviders.AddFirst(provider);
+
+        /// <summary>
+        /// Whether this options provider matches a given endpoint, for automatically selecting a provider based on what's being connected to.
+        /// </summary>
+        public virtual bool IsMatch(EndPoint endpoint) => false;
+
+        /// <summary>
+        /// Gets a provider for the given endpoints, falling back to <see cref="DefaultOptionsProvider"/> if nothing more specific is found.
+        /// </summary>
+        internal static Func<EndPointCollection, DefaultOptionsProvider> GetForEndpoints { get; } = (endpoints) =>
+        {
+            foreach (var endpoint in endpoints)
```

Comment:
I believe the outer loop should be on `KnownProviders`, and the inner loop on `endpoints`, so the first Provider in the list gets priority regardless of which endpoint matches

Reply:
I buy it, changed!